llvm-hs/llvm-hs
# Support LLVM 12.

Update `llvm-hs` from LLVM 9 to LLVM 12 (current top-of-tree).

List of changes:

```
- Attributes
  - Various minor changes, not listed in detail.
- Debug information
  - DIFlag: BlockByrefStruct and ArgumentNotModified removed.
  - DISubrange
    - New arguments: upperBound and stride.
    - Arguments {lowerBound,upperBound,stride} are generalized and may be
      DIVariable or DIExpression in addition to Constant Int64.
  - DIModule
    - New arguments: apiNotesFile and lineNo.
    - Removed arguments: isysroot.
- Instructions
  - New instructions: freeze.
  - Modified instructions
    - shufflevector: now takes mask as [Int32] instead of Constant.
- JIT
  - Some deprecated OrcJIT v1 APIs are now removed: VModuleKey, SymbolResolver.
  - Start updating APIs and tests using OrcJIT v2 APIs. Work-in-progress.
- Target
  - Target option flag changes, not listed in detail.
- Transforms
  - Some existing transforms were removed or merged into others.
    Some had options modified.

- Newly added APIs
  - Added Triple (target triple) API to llvm-hs-pure, along with string
    conversion and parsing.
  - Added debugging utilities: dumpMetadata.
- Tests
  - Textual LLVM IR changes
    - Function parameters are now printed with local identifiers (e.g. `%0`).
    - Instruction alignment arguments must now be explicit: 0 no longer seems
      like a valid default value.
  - Instrumentation tests
    - Update MemorySanitizer test to run only on supported platforms: matching
      the implementation at https://github.com/llvm/llvm-project/blob/release/12.x/llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp.
  - Optimization tests
    - Add a more robust "is module vectorized?" check for the SLPVectorization
      test.
```

Current status: all tests pass via `stack test`.

Todos:
- [x] Relevant comments to resolve: `{FIXME,TODO,NOTE}(llvm-12)`
- [x] Lots of code cleanup and API design finalization for OrcJIT v2.
- [x] Update installation instructions for `llvm-hs@llvm-12`.
- [x] Fix CI for `llvm-hs@llvm-12`.
